### PR TITLE
ARROW-17045: [C++] Reject trailing slashes on file path

### PR DESCRIPTION
**Pull request by wjones127**
BREAKING CHANGE: We had several different behaviors when passing in file paths with trailing slashes: LocalFileSystem would return IOError, S3 would trim off the trailing slash, and GCS would keep the trailing slash as part of the file name (later creating confusion as the file would be labelled a "directory" in list calls). This PR moves them all to the behavior of LocalFileSystem: return IOError.

The R filesystem bindings relied on the behavior provided by S3, since `FileSystem$path()` returns a `SubTreeFileSystem` as a convenient way to bundle a path and filesystem object and `SubTreeFileSystem$base_path` adds a trailing slash. To adapt to the C++ changes, the functions accepting `SubTreeFileSystem` as a path to a file now modified to trim the trailing slash before passing down to C++.

Here is an example of the differences in behavior between S3 and GCS:

```python
import pyarrow.fs
from pyarrow.fs import FileSelector
from datetime import timedelta

gcs = pyarrow.fs.GcsFileSystem(
    endpoint_override="localhost:9001",
    scheme="http",
    anonymous=True,
    retry_time_limit=timedelta(seconds=1),
)

gcs.create_dir("py_test")

# Writing to test.txt with and without slash produces a file and a directory!?
with gcs.open_output_stream("py_test/test.txt") as out_stream:
    out_stream.write(b"Hello world!")
with gcs.open_output_stream("py_test/test.txt/") as out_stream:
    out_stream.write(b"Hello world!")
gcs.get_file_info(FileSelector("py_test"))
# [<FileInfo for 'py_test/test.txt': type=FileType.File, size=12>, <FileInfo for 'py_test/test.txt': type=FileType.Directory>]

s3 = pyarrow.fs.S3FileSystem(
    access_key="minioadmin",
    secret_key="minioadmin",
    scheme="http",
    endpoint_override="localhost:9000",
    allow_bucket_creation=True,
    allow_bucket_deletion=True,
)

s3.create_dir("py-test")

# Writing to test.txt with and without slash writes to same file
with s3.open_output_stream("py-test/test.txt") as out_stream:
    out_stream.write(b"Hello world!")
with s3.open_output_stream("py-test/test.txt/") as out_stream:
    out_stream.write(b"Hello world!")
s3.get_file_info(FileSelector("py-test"))
# [<FileInfo for 'py-test/test.txt': type=FileType.File, size=12>]
```

### CODE REVIEW

**Comment by emkornfield** [comment]
Seems reasonable to me.   Is it possible to test this against a generated dataset to make sure nothing breaks there?

**Comment by wjones127** [comment]
> Seems reasonable to me. Is it possible to test this against a generated dataset to make sure nothing breaks there?

Do these tests from R seem sufficient? https://github.com/apache/arrow/blob/81af2912b5e088c83631df4cfd52e6d6df2878dc/r/tests/testthat/helper-filesystems.R#L130-L156

I pulled the changes from this PR into https://github.com/apache/arrow/pull/13542, so the above tests will run in CI there. (I can confirm they pass locally for GCS and S3.)

**Comment by jorisvandenbossche** [comment]
Just wondering: to what extent is this a corner case that just happens to be in our tests, or is there a practical value in allowing trailing slashes? 

Because for example also the LocalFileSystem does not allow trailing slashes in a similar example:

```
In [51]: from pyarrow.fs import LocalFileSystem

In [52]: local = LocalFileSystem()

In [53]: local.create_dir("py-test")

In [54]: with local.open_output_stream("py-test/test.txt/") as out_stream:
    ...:     out_stream.write(b"Hello world!")
...
IsADirectoryError: [Errno 21] Failed to open local file 'py-test/test.txt/'
Detail: [errno 21] Is a directory
```

(which is an error that makes sense to me, also Python's `open` will complain that it is a directory and not a file in the equivalent example)

**Comment by pitrou** [comment]
I agree it seems better to disallow trailing slashes in filenames rather than silently dropping them.

**Comment by wjones127** [comment]
> Just wondering: to what extent is this a corner case that just happens to be in our tests, or is there a practical value in allowing trailing slashes?

I'm not sure yet what else relies on that. In R, it's just that we pass `fs$path("some/path")` to functions like `write_parquet()`. `fs$path()` returns a `SubTreeFilesystem`, which I think is just used as a convenient way to pass a path (as the base path) and a filesystem in one object. But as a side effect of using `base_path` to transmit the path, we add a trailing slash. Using this method fails for `LocalFileSystem` though:

``` r
library(arrow)
#> 
#> Attaching package: 'arrow'
#> The following object is masked from 'package:utils':
#> 
#>     timestamp

example_data <- arrow_table(x = Array$create(c(1, 2, 3)))

fs <- LocalFileSystem$create()
write_parquet(example_data, fs$path("test.parquet"))
#> Error: IOError: Failed to open local file 'test.parquet/'
#> /Users/willjones/Documents/arrows/arrow/cpp/src/arrow/filesystem/localfs.cc:442  ::arrow::internal::FileOpenWritable(fn, write_only, truncate, append). Detail: [errno 2] No such file or directory
```

<sup>Created on 2022-07-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

cc @nealrichardson in case he has any input.

I will see about changing GCS to reject file paths that end with slashes. Should I leave S3 alone? Or should it also reject them?

**Comment by pitrou** [comment]
Hmm, `SubTreeFilesystem` is meant to take a subdirectory parameter, you are not expected to use it to pass a filename.

**Comment by pitrou** [comment]
> Should I leave S3 alone? Or should it also reject them?

It would be fine with me to remove them, but if you hit too many regressions then no need to sweat over it either :-)

**Comment by nealrichardson** [comment]
> Hmm, `SubTreeFilesystem` is meant to take a subdirectory parameter, you are not expected to use it to pass a filename.

Historical context: ARROW-10254, which points to https://github.com/apache/arrow/pull/8351#discussion_r501252247

We could have the single-file writers (perhaps the readers too) in R prune a possible trailing slash from filenames, if this is the only source of the issue. That logic is pretty well encapsulated in `make_readable_file()` and `make_output_stream()` so it should be feasible to do. Shouldn't be a concern for write_dataset/open_dataset since they point at directories.

File: cpp/src/arrow/filesystem/path_util.cc [hunk 1]
```diff
@@ -139,6 +140,13 @@ util::string_view RemoveLeadingSlash(util::string_view key) {
   return key;
 }
 
+Status AssertNoTrailingSlash(const std::string& key) {
+  if (key.back() == '/') {
+    return NotAFile(key);
```
Comment (wjones127) [inline cpp/src/arrow/filesystem/path_util.cc:145]: I would take `util::string_view` instead but `NotAFile` only accepts `const std::string&`. Would it be alright if I moved "arrow/filesystem/util_internal.h" to use `util::string_view`?

Comment (pitrou) [reply cpp/src/arrow/filesystem/path_util.cc:145]: Definitely ok!

File: r/R/io.R [hunk 2]
```diff
@@ -241,7 +241,7 @@ mmap_open <- function(path, mode = c("read", "write", "readwrite")) {
 make_readable_file <- function(file, mmap = TRUE, compression = NULL, filesystem = NULL) {
   if (inherits(file, "SubTreeFileSystem")) {
     filesystem <- file$base_fs
-    file <- file$base_path
+    file <- gsub("/$", "", file$base_path)
```
Comment (nealrichardson) [inline r/R/io.R:244]: nit: no need for global substitution

```suggestion
    file <- sub("/$", "", file$base_path)
```

Comment (nealrichardson) [reply r/R/io.R:244]: This is also worth leaving a code comment to explain why

File: r/R/io.R [hunk 3]
```diff
@@ -241,7 +241,9 @@ mmap_open <- function(path, mode = c("read", "write", "readwrite")) {
 make_readable_file <- function(file, mmap = TRUE, compression = NULL, filesystem = NULL) {
   if (inherits(file, "SubTreeFileSystem")) {
     filesystem <- file$base_fs
-    file <- file$base_path
+    # SubTreeFileSystem adds a slash to base_path, but filesystems will reject file names
+    # with trailing slashes, so we need to remove it here.
+    file <- sub("/$", "", file$base_path)
```
Comment (pitrou) [inline r/R/io.R:246]: I'm probably misreading this, but is this treating any `SubTreeFileSystem` as pointing to a local file path?

Comment (wjones127) [reply r/R/io.R:246]: By "filesystems" I meant the Arrow FileSystem classes, not the local file system. Does that clarify your confusion? Or are you asking something else?

**Review by nealrichardson — APPROVED** [review]
R side LGTM, thanks!

File: r/R/io.R (hunk 3 above)
Comment (pitrou) [reply r/R/io.R:246]: No, I am asking something else. It seems that this code is replacing `file` (a FileSystem instance) with a file path, is that right? And below, the file path `file` will be treated as a local filesystem path?

Comment (wjones127) [reply r/R/io.R:246]: But the path should be treated as correct filesystem since we extracted that out in the line before.

Comment (pitrou) [reply r/R/io.R:246]: Oh, I see, I had missed the `filesystem <- file$base_fs`. So I guess my last question is: why are we calling `make_readable_file` with a `SubTreeFileSystem` as the first argument?

Comment (wjones127) [reply r/R/io.R:246]: The expectation is that users will call `FileSystem$path()` to specify the location to write/read to:

```R
fs <- S3FileSystem$create()
write_parquet(my_tab, fs$path("my/path/to"))
```

`fs$path` returns a `SubTreeFileSystem` as a convenient way to bundle a path and filesystem object in a single object. See discussion: https://github.com/apache/arrow/pull/8351#discussion_r501252247

Comment (pitrou) [reply r/R/io.R:246]: Ah. That's unfortunate, but that's not this PR's business. Thanks for the details!

**Review by pitrou — APPROVED** [review]
+1 from me, this is a reasonable straightening of the API. Thank you @wjones127 for doing this!

**Comment by wjones127** [comment]
Oh no! Those do look related @jorisvandenbossche. Is HDFS not in our usual set of CI tests?

**Pull request merged**